refactor(compiler): add `i18nPreserveWhitespaceForLegacyExtraction`

dgp1130 opened the pull request:
CARETAKER NOTE: I have passing TGPs with this flag both enabled and disabled. I just used Fig so it didn't propagate the status to the GitHub check cleanly.

This configures whether or not to preserve whitespace content when extracting messages from Angular templates in the legacy (View Engine) extraction pipeline. It defaults to `true` to maintain backwards compatibility, however `false` is the ideal behavior since it makes translations less sensitive to unimportant whitespace changes.

The change to `i18n_parser.ts` is really a bug fix. The localization process requires that the number of tokens in a message stays constant between i18n passes so it can reuse source spans. Therefore even empty text tokens need to be retained in order to maintain that invariant. However this bug fix does affect message IDs, and changing that could invalidate existing translations, therefore it is also gated by the same flag to maintain backwards compatibility.

There are two other issues I discovered while adding this flag. First, Angular template expressions are factored in to the message ID calculation. Meaning changing from `Hello, {{ name }}` to `Hello, {{ firstName }}` will trigger retranslatation. That _almost_ sounds correct, except that all expressions are replaced with something like `{INTERPOLATION}` unless `// ph(name = "...")` is used, so changing the expression would have no visible effect to translators anyways. Therefore, disabling `i18nPreserveWhitespaceForLegacyExtraction` removes placeholders from the calculation.

Second, ICU templates are broken up into multiple messages based on whitespace. So `<div i18n>{...}</div>` and `<div i18n> {...} </div>` generate one and two `$localize` messages apiece. Changing this would require updating the generated runtime code, which is out of scope for the time being here.

dgp1130 committed:
refactor(compiler): add `i18nPreserveWhitespaceForLegacyExtraction`

This configures whether or not to preserve whitespace content when extracting messages from Angular templates in the legacy (View Engine) extraction pipeline.

This includes several bug fixes which unfortunately cannot be landed without changing message IDs in a breaking fashion and are necessary to properly trim whitespace. Instead these bug fixes are included only when the new flag is disabled.

AndrewKushnir commented on packages/compiler-cli/src/transformers/i18n.ts:
```diff
@@ -57,7 +57,9 @@ export function i18nSerialize(
     case 'xmb':
-      serializer = new Xmb();
+      serializer = new Xmb(
+        options.i18nPreserveWhitespaceForLegacyExtraction /* preservePlaceholders */,
```
nit: we can add a quick comment into the code here to describe a connection between preserving whitespaces and preserving placeholders.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -90,8 +92,30 @@ export class WhitespaceVisitor implements html.Visitor {
       );
+
+      // Fully trim message when significant whitespace is not preserved.
+      if (!this.preserveSignificantWhitespace && tokens.length) {
```
nit (for more explicitness, this is a more common pattern in most recent compiler code):
```suggestion
      if (!this.preserveSignificantWhitespace && tokens.length > 0) {
```

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -135,6 +159,13 @@ function createWhitespaceProcessedTextToken({type, parts, sourceSpan}: TextToken
+  {type, parts, sourceSpan}: TextToken,
+  transform: (parts: string) => string,
+): TextToken {
+  return {type, parts: [transform(parts[0])], sourceSpan};
```
Should we retain all `parts` here (instead of trimming the first one and storing just that token)?

dgp1130 replied:
`TextToken` is defined as a type having only one element in the array and the type system should prevent misuse of that. Added a comment to make this more clear.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -90,8 +92,30 @@ export class WhitespaceVisitor implements html.Visitor {
       // Process the whitespace of the value of this Text node
-      const value = processWhitespace(text.value);
+      const processed = processWhitespace(text.value);
+      const value = this.preserveSignificantWhitespace ? processed : processed.trim();
```
nit: it'd be great to add a comment here to explain why we trim whitespaces from the final message text.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -48,13 +48,21 @@ export function replaceNgsp(value: string): string {
       // don't descent into elements where we need to preserve whitespaces
       // but still visit all attributes to eliminate one used as a market to preserve WS
-      return new html.Element(
+      const el = new html.Element(
```
nit:
```suggestion
      const newElement = new html.Element(
```

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -90,9 +102,37 @@ export class WhitespaceVisitor implements html.Visitor {
+        // The first token should only call `.trimStart()` and the last token
+        // should only call `.trimEnd()`, but there might be only one token which
+        // needs to call both.
+        const firstToken = tokens[0]!;
```
I'd propose refactoring this code into one or two smaller functions, which we can invoke here.

AndrewKushnir commented on packages/compiler/src/i18n/message_bundle.ts:
```diff
@@ -41,11 +44,22 @@ export class MessageBundle {
+    const rootNodes = this._preserveWhitespace
+      ? htmlParserResult.rootNodes
+      : html.visitAll(
+          new WhitespaceVisitor(false /* preserveSignificantWhitespace */),
```
It'd be great to add a comment here as well explaining why we don't preserve significant whitespaces.

dgp1130 replied:
There's a comment above `rootNodes` which explains this.

AndrewKushnir reviewed (approved):
@dgp1130 looks great, thanks for the updates 👍 I've left a few minor comments, but none of them are blocking.

As a followup (in a separate PR), it might be great to put together a markdown file (e.g. `packages/compiler/src/i18n/EXTRACTION.md`) and describe how the extraction process depends on whitespaces in messages and how different subsystems are configured to deal with it (e.g. ViewEngine-based extraction, code generation with goog.getMsg, etc). You have this information already in different comments in the code (which is very helpful!) and a document may help provide an overall terminology (e.g. what a "significant whitespace" means), how this works from high-level and we can also use the doc as a reference in comments to the code that you've added. This document will help us in the future to restore the context once we decide to perform further changes to this code.

Thank you.

thePunderWoman reviewed (approved):
reviewed-for: public-api

dgp1130 committed:
refactor(compiler): ensure context is always provided for `WhitespaceVisitor`

When disabling `i18nPreserveSignificantWhitespaceForLegacyExtraction` I was looking at a test case with ICU messages containing leading and trailing whitespace:

```angular
<div i18n>
  {apples, plural, =other {I have many apples.}}
</div>
```

This would historically generate two messages:

```javascript
const MSG_TMP = goog.getMsg('{apples, plural, =other {I have many apples.}}');
const MSG_FOO = goog.getMsg(' {$ICU} ', { 'ICU': MSG_TMP });
```

But I found that I was getting just one message:

```javascript
const MSG_TMP = goog.getMsg(' {apples, plural, =other {I have many apples.}} ');
```

This is arguably an improvement, but changed the messages and message IDs, which isn't desirable with this option. I eventually traced this back to the `isIcu` initialization in [`i18n_parser.ts`](/packages/compiler/src/i18n/i18n_parser.ts):

```typescript
const context: I18nMessageVisitorContext = {
  isIcu: nodes.length == 1 && nodes[0] instanceof html.Expansion,
  // ...
};
```

[`_I18nVisitor.prototype.visitExpansion`](/packages/compiler/src/i18n/i18n_parser.ts) uses this to decide whether or not to generate a sub-message for a given ICU expansion:

```typescript
if (context.isIcu || context.icuDepth > 0) {
  // Returns an ICU node when:
  // - the message (vs a part of the message) is an ICU message, or
  // - the ICU message is nested.
  const expPh = context.placeholderRegistry.getUniquePlaceholder(`VAR_${icu.type}`);
  i18nIcu.expressionPlaceholder = expPh;
  context.placeholderToContent[expPh] = {
    text: icu.switchValue,
    sourceSpan: icu.switchValueSourceSpan,
  };
  return context.visitNodeFn(icu, i18nIcu);
}

// Else returns a placeholder
// ICU placeholders should not be replaced with their original content but with the their
// translations.
// TODO(vicb): add a html.Node -> i18n.Message cache to avoid having to re-create the msg
const phName = context.placeholderRegistry.getPlaceholderName('ICU', icu.sourceSpan.toString());
context.placeholderToMessage[phName] = this.toI18nMessage([icu], '', '', '', undefined);
const node = new i18n.IcuPlaceholder(i18nIcu, phName, icu.sourceSpan);
return context.visitNodeFn(icu, node);
```

Note that `isIcu` is the key condition between these two cases and depends on whether or not the ICU expansion has any siblings. The introduction of `WhitespaceVisitor` to `I18nMetaVisitor` trims insignificant whitespace, including empty text nodes not adjacent to an ICU expansion (from [`WhitespaceVisitor.prototype.visitText`](/packages/compiler/src/ml_parser/html_whitespaces.ts)):

```typescript
const isNotBlank = text.value.match(NO_WS_REGEXP);
const hasExpansionSibling =
  context && (context.prev instanceof html.Expansion || context.next instanceof html.Expansion);

if (isNotBlank || hasExpansionSibling) {
  // Transform node by trimming it...
  return trimmedNode;
}

return null; // Drop node which is empty and has no ICU expansion sibling.
```

`hasExpansionSibling` was intended to retain empty text nodes leading or trailing an ICU expansion, however `context` was `undefined`, so this check failed and the leading / trailing text nodes were dropped. This resulted in trimming the ICU text by dropping the leading / trailing whitespace nodes. Having only a single ICU expansion with no leading / trailing text nodes caused `_I18nVisitor` to initialize `isIcu` incorrectly and caused it to generate one message instead of two.

`WhitespaceVisitor` is supposed to get this context from `visitAllWithSiblings`. So the fix here is to make sure `WhitespaceVisitor` is always visited via this function which provides the required context. I updated all usage sites to make sure this context is use consistently and implemented the `WhitespaceVisitor.prototype.visit` method to throw when the context is missing to make sure we don't encounter a similar mistake in the future.

Unfortunately this broke one compliance test. Specifically the [`icu_logic/icu_only.js`](/home/douglasparker/Source/ng/packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/icu_logic/icu_only.js) test which changed from generating:

```javascript
function MyComponent_Template(rf, ctx) {
  if (rf & 1) {
    i0.ɵɵi18n(0, 0);
  }
  // ...
}
```

To now generating:

```javascript
function MyComponent_Template(rf, ctx) {
  if (rf & 1) {
    i0.ɵɵtext(0, " ");
    i0.ɵɵi18n(1, 0);
    i0.ɵɵtext(2, "\n");
  }
  // ...
}
```

This test uses the default value `preserveWhitespaces: false` (`i18nPreserveSignificantWhitespaceForLegacyExtraction` should not affect compiled JS output, we already retain significant whitespace there). So what this indicates to me is that ICU logic is already broken because it's not preserving significant whitespace in this case. My change is probably a bug fix, but one which would affect the compiled runtime, which is not in scope here. The root cause is because using `visitAllWithSiblings` everywhere means the context is retained correctly in this case and the whitespace is leading/trailing an ICU message, therefore it is retained per the logic of `WhitespaceVisitor.prototype.visitText` I mentioned eariler.

To address this, I left one usage of `WhitespaceVisitor` using `html.visitAll` instead of `visitAllWithSiblings` to retain this bug. I has to lossen the assertion I put in `WhitespaceVisitor.prototype.visit` to make this possible, but it should still throw by default when misused, which is the important part.

dgp1130 commented:
TGP: http://test/OCL:645226727:BASE:654048129:1721410070694:c257832d

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -278,6 +280,7 @@ export class ComponentDecoratorHandler
     usePoisonedData: boolean;
     enableBlockSyntax: boolean;
     enableLetSyntax: boolean;
+    preserveSignificantWhitespace?: boolean;
```
In the compiler we try to avoid optional parameters.
```suggestion
    preserveSignificantWhitespace: boolean;
```

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -1730,6 +1732,23 @@ ${allowedCategoryLabels.join('\n')}
 }
 
+/** Checks i18n options for internal consistency and emits diagnostics for any discovered issues. */
+function* verifyI18nOptions(options: NgCompilerOptions): Generator<ts.Diagnostic, void, void> {
```
Out of curiosity, why a generator and not an `Array<ts.Diagnostic>`?

dgp1130 replied:
Removed this error check altogether.

alxhub commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -32,15 +32,18 @@ export function computeDigest(message: i18n.Message): string {
  */
-export function decimalDigest(message: i18n.Message): string {
-  return message.id || computeDecimalDigest(message);
+export function decimalDigest(message: i18n.Message, preservePlaceholders?: boolean): string {
```
```suggestion
export function decimalDigest(message: i18n.Message, preservePlaceholders: boolean): string {
```

alxhub commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -100,12 +103,23 @@ export function serializeNodes(nodes: i18n.Node[]): string[] {
-class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
-  override visitIcu(icu: i18n.Icu, context: any): any {
+class _SerializerIgnoreExpVisitor extends _SerializerVisitor {
+  constructor(private readonly preservePlaceholders = true) {
```
```suggestion
  constructor(private readonly preservePlaceholders) {
```

alxhub commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -37,8 +39,14 @@ export interface I18nMessageFactory {
 export function createI18nMessageFactory(
   interpolationConfig: InterpolationConfig,
   containerBlocks: Set<string>,
+  retainEmptyTokens?: boolean,
```
```suggestion
  retainEmptyTokens: boolean,
```

alxhub commented on packages/compiler/src/i18n/serializers/xmb.ts:
```diff
@@ -48,6 +48,10 @@ const _DOCTYPE = `<!ELEMENT messagebundle (msg)*>
 <!ELEMENT ex (#PCDATA)>`;
 
 export class Xmb extends Serializer {
+  constructor(private readonly preservePlaceholders?: boolean) {
```
```suggestion
  constructor(private readonly preservePlaceholders: boolean) {
```
(unless this needs to be optional for compatibility with external usages)

dgp1130 replied:
This is exported from `@angular/compiler`, so I think it qualifies as public API and can't be changed right now. I left it optional, but did update all internal usages to explicitly pass this value in.

alxhub reviewed (approved):
LGTM with some nits on optional arguments which are discouraged/avoided in compiler code.

dgp1130 reviewed (commented):
I just removed the error check preventing `preserveWhitespaces = true` from being used with `i18nPreserveWhitespaceForLegacyExtraction = false`.

It sounds reasonable on paper that preserving insignificant whitespace requires preserving significant whitespace, but that's not really what this new flag does. The new flag should really be called: `i18nPreserveSignificantWhitespaceAndPlaceholderWhitespaceForLegacyExtractionButNotMergeAndAlsoFixSomeUnrelatedBugsThatBecomeAProblemRightNow`. So that doesn't _exactly_ map to `preserveSignificantWhitespace`. Specifically, `preserveWhitespace` controls extraction and merging behavior for insignificant whitespace, but this new flag only controls extraction behavior for significant whitespace. So it is valid to set `preserveWhitespace: true, i18nPreserveWhitespaceForLegacyExtraction: false`, no matter how weird that sounds.

I'm still running some tests in g3, but will hopefully get back to Alex's comments soon.

dgp1130 commented:
I went through a sizeable workflow of extract -> format -> re-extract -> compare to verify that message IDs are indeed durable to whitespace changes with this flag configured. I found ~15 messages which changed IDs, and here's a summary of what caused those changes to document the known limitations of this flag:

*   Duplicate placeholders with expressions different only in whitespace will generate different names.
    ```html
    <div i18n>My name is {{ name }}. I love my name {{            name           }}.</div>
    ```
    It's arguably wrong for Angular to try and reuse a placeholder for multiple expressions at all because even if the expressions are exactly identical, we can't guarantee they will return the value same value `{{ append(name) }}` could return a different values on the second execution if `append` is not pure. So I'm not particularly worried about fixing this one.
*   Spacing and line breaks in placeholder names are significant. This is an existing bug, but realistically doesn't come up a whole lot, and most devs likely wouldn't want to check this in anyways, so I don't think it's super important to fix.
    ```html
    <div i18n>Hello, {{ name // ph(name = "Name of
        the user
    ")</div>
    ```
*    [A few tags](https://github.com/angular/angular/blob/3b0dca75d6dab6039253edd00c436715775bd0dd/packages/compiler/src/ml_parser/html_whitespaces.ts#L16) (notably `<pre>`) do not get trimmed because all whitespace is significant by default for them. This isn't a bug at all and is WAI.
*   `ngPreserveWhitespace` messages do not get trimmed. I think this is correct behavior and would be necessary for the case of `<div style="white-space: pre;">    Some message    </div>`.
*   Spacing and line breaks in message descriptions are significant.
    ```html
    <div i18n="Long message
            description which line wraps">
        Message
    </div>
    ```
    
    Prettier doesn't format this today so I don't think it's an immediate issue, but hypothetically it could format this in the future. I debated with myself about this for a while, but ultimately I'm not convinced we necessarily should trim this text as there's no implied HTML whitespace semantics in an attribute like this and devs could take advantage of that to write detailed, multi-line descriptions:
    
    ```html
    <div i18n="
    Long message with a list:
    * First talk about this.
    * Then talk about this.
    * Finally end about this.
    ">
        Message
    </div>
    ```
    
    I'm not convinced introducing HTML's whitespace semantics here is necessarily a good idea, so I'm inclined to leave this alone for the time-being until it becomes an issue or we decide to introduce whitespace semantics here. We can then have a separate discussion on exactly _which_ semantics would be appropriate for this context.
*   Line wrapping control flow structures are significant. Compare the following two examples only different in whitespace:
    ```html
    <div i18n>
        Foo@if (cond) {Bar}
    </div>
    ```

    ```html
    <div i18n>
        Foo
        @if (cond) {
            Bar
        }
    </div>
    ```

    The former will use "FooBar" while the later will have a space "Foo Bar". I think this is WAI from an i18n perspective (rendered HTML is the same), it's just not safe to make this kind of transformation in general.

Next steps are to address the active comments on this PR, rerun tests to verify everything is in a good state, and then hopefully merge.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -48,24 +48,38 @@ export function replaceNgsp(value: string): string {
  * and might be changed to "on" by default.
+ *
+ * If `originalNodeMap` is provided, the transformed nodes will be mapped back to their original
+ * inputs. Any output nodes not in the map were not transformed.
```
nit: I'd propose adding a bit more info on what `originalNodeMap` represents and why we need it (as an additional context for future maintenance).

AndrewKushnir commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -103,15 +145,40 @@ export class WhitespaceVisitor implements html.Visitor {
+        expansion.switchValueSourceSpan,
+        expansion.i18n,
+      );
+    } finally {
+      this.inExpansion = false;
```
I'm curious if that would work correctly for nested ICUs, e.g.:
```
{select, emailCount,
  =0 { zero in inbox {select, deletedCount, =0 {no removed emails} other {has removed emails}} }
  other {    Message with some whitespaces   }
}
```
Not sure if this is significant edge case though.

dgp1130 replied:
Good call out, updated to use an increment `icuExpansionDepth` counter over a single `boolean` so we can track this correctly.

I tried to reproduce this example, but actually wasn't really able to. The second ICU doesn't get formatted because my new logic is gated on `preserveSignificantWhitespace === false` but this case only happens when `preserveSignificantWhitespace === true`. However it does trigger some trimming of `text.value`, so this is still a good fix to make.

AndrewKushnir reviewed (approved):
Looks great, just left a couple comments.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

dgp1130 commented:
FYI: I have passing TGPs with this flag both enabled and disabled. I just used Fig so it didn't propagate the status to the GitHub check cleanly.

alxhub commented:
This PR was merged into the repository by commit 5a0ff41f75fd6d58d73baf11315c012131d9ec82.

The changes were merged into the following branches: main, 18.2.x
